metaflow/plugins/aws/awslambda/lambda_decorator.py

```diff
@@ -0,0 +1,208 @@
+import sys
+
+from metaflow import R
+from metaflow.datastore.datastore import TransformableObject
+from metaflow.decorators import Decorator, StepDecorator
+from metaflow.exception import MetaflowException
+from metaflow.plugins.aws.awslambda.lambda_runner import LAMBDA_MAX_MEMORY_MB
+from metaflow.plugins.aws.batch.batch_decorator import (
+    DEFAULT_TIMEOUT_SECONDS,
+    BatchDecorator,
+    ResourcesDecorator,
+)
+from metaflow.plugins.aws.utils import sync_metadata_to_s3
+from metaflow.plugins.conda.conda_step_decorator import CondaStepDecorator
+from metaflow.plugins.timeout_decorator import get_run_time_limit_for_task
+
+if sys.version_info > (3, 0):
+    from typing import TYPE_CHECKING, Dict, List, Optional
+
+    if TYPE_CHECKING:
+        from metaflow.runtime import CLIArgs
+        from metaflow.graph import FlowGraph
+        from metaflow.package import MetaflowPackage
+        from metaflow.flowspec import FlowSpec
+        from metaflow.datastore import MetaflowDataStore
+        from metaflow.metaflow_environment import MetaflowEnvironment
+        from metaflow.event_logger import EventLogger
+
+from metaflow.metaflow_config import (
+    LAMBDA_IMAGE_URI,
+    LAMBDA_ROLE_ARN,
+)
+
+DEFAULT_TIMEOUT_SECONDS = 600
+
+
+class LambdaException(MetaflowException):
+    headline = "AWS Lambda error"
+
+
+class LambdaDecorator(StepDecorator):
+    """
+    The decorator is called @awslambda to avoid confusion, since "lambda" is a reserved
+    word in Python.
+    """
+
+    name = "awslambda"
+    defaults = {
+        "image": None,
+        "force_fetch_image": False,
+    }
+
+    package_url = None  # type: Optional[str]
+    package_sha = None  # type: Optional[str]
+    resource_attributes = {}  # type: Dict[str, str]
+    run_time_limit = None
+    ds_root = None  # type: Optional[str]
+
+    def step_init(
+        self,
+        flow,  # type: "FlowSpec"
+        graph,  # type : "FlowGraph"
+        step,  # type: str
+        decos,  # type: List[Decorator]
+        environment,  # type: "MetaflowEnvironment"
+        datastore,  # type: "MetaflowDataStore"
+        logger,  # type: "EventLogger"
+    ):
+        if datastore.TYPE != "s3":
+            raise LambdaException("The *@awslambda* decorator requires --datastore=s3.")
+
+        self.logger = logger
+        self.environment = environment
+        self.step = step
+        for deco in decos:
+            if isinstance(deco, CondaStepDecorator):
+                raise LambdaException(
+                    "*@awslambda* doesn't support using @conda. "
+                    "We recommend using custom docker images to manage"
+                    " dependencies instead"
+                )
+
+            if isinstance(deco, BatchDecorator):
+                raise LambdaException(
+                    "You can't use @batch and *@awslambda* at the same time"
```

Comment:
We need to move this check outside this routine and check for the validity of the flow before the flow starts executing.

Reply:
I agree but what's a good place for this? At first I looked at `FlowLinter` but it doesn't support plugins today

Reply:
I will commit a `PluginLinter` to the codebase to help with this.